stream-chat-android-state/src/main/java/io/getstream/chat/android/offline/plugin/logic/channel/internal/TypingEventPruner.kt

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.offline.plugin.logic.channel.internal
+
+import androidx.annotation.VisibleForTesting
+import io.getstream.chat.android.client.events.TypingStartEvent
+import io.getstream.chat.android.client.models.TypingEvent
+import io.getstream.chat.android.core.internal.InternalStreamChatApi
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+
+/**
+ * A class used to keep track of typing events
+ * and remove stale ones that were not explicitly stopped by the sender
+ * due to technical difficulties. (e.g. process death, loss of Internet
+ * connection)
+ *
+ * @property coroutineScope The scope used to launch jobs.
+ * @param delayTimeMs The period of time it takes before an individual stale
+ * typing event is removed.
+ */
+internal class TypingEventPruner(
+    private val coroutineScope: CoroutineScope,
+    private val delayTimeMs: Long = DEFAULT_DELAY_TIME_MS,
+) {
+
+    /**
+     * The id of the channel for which the typing
+     * events are being pruned.
+     */
+    private lateinit var channelId: String
+
+    /**
+     * Triggered when typing events have been updated.
+     *
+     * Use [setOnTypingEventsUpdatedListener] to set a listener externally.
+     */
+    private var onUpdated: (
+        rawTypingEvents: Map<String, TypingStartEvent>,
+        typingEvent: TypingEvent,
+    ) -> Unit = { _, _ -> }
+
+    /**
+     * A mutable map of typing events.
+     *
+     * Each event is capable of timed death using [TimedTypingStartEvent.removeTypingEvent]
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    val typingEvents = mutableMapOf<String, TimedTypingStartEvent>()
+
+    /**
+     * Initializes the pruner by passing in and setting
+     * the ID of the channel that the pruner is responsible for.
+     *
+     * @param channelId The ID of the channel the pruner is responsible for.
+     */
+    fun initialize(channelId: String) {
+        this.channelId = channelId
+    }
```

Comment:
`channelId` looks like it will be always the same (At least `initialize` name told me it should only be called on the initialization process), Why can't we inject it by constructor? :thinking:  It looks as a simple string that we should have before the `TypingEventPruner` is created

Reply:
You are right, this is a leftover from a previous strategy where I thought initialization might look different.
It should be optimized.

Reply:
Done :smile: